packages/services-context/src/index.ts

```diff
@@ -99,10 +101,14 @@ export class Services implements ServicesInterface {
       sctParams: { epochDuration },
     } = params;
 
+    const registryClient = new ChainRegistryClient();
+    const { stakingAssetId, assetById, numeraires } = await registryClient.get(chainId);
```

Comment:
Grab from IndexedDB if possible

Reply:
I'm not sure I'm understanding your idea correctly. 
The thing is that I get the `assetById` here just to pass to indexed-db for initialization. 
On initialization indexed-db will pre-populate the asset table with these values. 
And yes, this will happen every time indexed-db is initialized. 

Regarding `numeraires` and `stakingAssetId`, I can create a table for them in indexed-db, but to populate that table with data I still have to go to registry, I don't really understand what problem this solves, at the same time we have to think about the fact that `numeraires` or `stakingAssetId` in indexed-db might be considered outdated

Reply:
I might be misunderstanding something here, but let me explain what I'm trying to suggest and see if it sounds right to you.

So, at the moment, we generally avoid making calls to an RPC node (or any other server, like GitHub) whenever possible — _except_ for in the block processor. In the block processor, we make calls to the RPC node whenever we need to, since that doesn't reveal anything about the user's intentions. Everyone who uses the Prax extension uses the block processor and thus makes the exact same RPC/API calls.

But if we use the `ChainRegistryClient` in both minifront and in the extension's RPC implementations, those calls will fire on an as-needed basis, which may leak information about the user's intentions.

Thus, my argument is that we should fetch data from the registry in the block processor [here](https://github.com/penumbra-zone/web/blob/main/packages/query/src/block-processor.ts#L322). That way, we're guaranteed to have an up-to-date set of asset metadata at any given time, since we'll always know about every asset type that's been traded/etc. on chain.

Let me know if my understanding is correct — perhaps my thinking doesn't apply to IBC, since we'll need to know about tokens that _can_ be traded in/out of Penumbra, even if they haven't been yet?

Reply:
> That way, we're guaranteed to have an up-to-date set of asset metadata at any given time, since we'll always know about every asset type that's been traded/etc. on chain.
the `saveAndReturnMetadata()` function is called only for those assets that occur on the user's balance. This does not exactly guarantee that we will have a local copy of the metadata of all assets in the indexed-db


I can see some disadvantages to using a local copy of the asset registry 
- we do not know when the local copy was last updated and whether it contains an up-to-date list of assets
- In addition to saving, we will probably have to make a mechanism for updating the metadata and deleting the metadata. (I've presented a case where the registry initially had the wrong metadata but was corrected afterwards, but the local copies will never be changed).  

Also, it seems to me that there is no privacy issue in getting data from github because in fact we are reading all datadata from the registry (ibc, rpc, all assets metadata) on every call (one json file is being read) and it literally does not reveal any of our intentions. 
Also, as far as I understand, Gabe has created a mechanism to cache the registry, meaning that the data will be read only once, and the next calls will use the cache. 

And considering that we already use the registry to get ibc channel data, nothing changes from the point of view of the github worker who is watching us

Reply:
`chainId` should not change often, and if it does our state is invalid - we query app parameters in order to verify it has not changed. i believe `stakingTokenId` should not change unless `chainId` changes.

there is no explicit `stakingTokenId` in app parameters, but i believe staking tokens should be the only asset present at the genesis block, so could detect it without using the registry client.

staking token id is generated in core repo from known values https://github.com/penumbra-zone/penumbra/blob/main/crates/core/asset/src/asset/cache.rs#L28

we may be able to expose this via wasm. https://github.com/penumbra-zone/penumbra/blob/main/crates/core/asset/src/asset/id.rs#L156

notably LP uses the same hash and technique, which is exposed via wasm https://github.com/penumbra-zone/penumbra//blob/main/crates/core/component/dex/src/lp/position.rs#L103

Reply:
> we may be able to expose this via wasm

Yes, but it doesn't improve anything because we would still need to get the metadata (from the registry or indexed-db) and we would still need to call the registry for other things (assets, numeraires) 
Also, we probably don't want minifront to use wasm, and we need stakingTokenId there too

Reply:
OK, gotcha. Well, my instinct is to say that it probably isn't a big concern if these requests are being sent to GitHub, since the RPC node has no access to GitHub's internal server access logs. But I'm still a little wary of approving it — this is probably a question for @grod220 since it's sort of a product issue.

Reply:
> So, at the moment, we generally avoid making calls to an RPC node (or any other server, like GitHub) whenever possible — except for in the block processor.

I think the principle is something like:
- Do not make a request to node when we can locally serve the result from indexeddb
- If making requests to node, do our best to ensure it doesn't leak privacy (aka query node for whole block instead of specific tx’s).
- If there’s no way to obfuscate, make it opt in (e.g. swap estimations are triggered via button with a privacy notice)

In this case, querying the registry at initialization does not reveal any specific intentions besides wanting to use the system generally. I don’t see an issue with this.

However, as has been suggested in the review, minifront should be requesting asset denom data via rpc calls and not from github directly.

> the saveAndReturnMetadata() function is called only for those assets that occur on the user's balance. This does not exactly guarantee that we will have a local copy of the metadata of all assets in the indexed-db

I agree, we should be able to utilize the registry to pre-populate at initialization and the block processor to add to the assets table if it discovers new ones.

> I can see some disadvantages to using a local copy of the asset registry
> - we do not know when the local copy was last updated and whether it contains an up-to-date list of assets

Given this happens at service worker initialization, indexeddb should pretty much always have the latest from the registry.

> In addition to saving, we will probably have to make a mechanism for updating the metadata and deleting the metadata.

Saving just overwrites the old, so this is probably fine. Not sure the use-case for deleting, but can revisit that later.